Change pedersen_hash_exp_window_size to 16.

porcuquine opened the pull request:
Based on analaysis in https://github.com/filecoin-project/rust-fil-proofs/issues/697 and comments in https://github.com/filecoin-project/rust-fil-proofs/issues/736, it seems clear that bumping the `pedersen_hash_exp_window_size` from 8 to 16 as the default is a safe change worthwhile making no matter what. This accomplishes a ~1.7X speedup of pedersen hashes outside of circuits, with an apparent measured impact of only ~50K on max memory usage.

Therefore, rather than sit on this optimization which will be useful for other efforts and general momentum, let's change the default now but still continue with https://github.com/filecoin-project/rust-fil-proofs/issues/736 so we can further refine the optimal setting and allow a greater range of options for when we start trying to derive 'final' configurations.

porcuquine committed:
Change pedersen_hash_exp_window_size to 16.

porcuquine commented:
A wrinkle: building on CI is causing circuit and redjubjub tests to timeout: https://circleci.com/gh/filecoin-project/sapling-crypto/35.

So… it may be that we *do* need configurability and to keep 8 as the default to avoid this problem. This would allow us to select a different window size when hashing outside of circuits.

We should add benchmarks for circuit performance (speed/memory) and independently settle on an optimal value there if it turns out to (as it seems) to produce different effects in that context.

```
test circuit::blake2s::test::test_blake2s ... test circuit::blake2s::test::test_blake2s has been running for over 60 seconds
test circuit::ecc::test::test_conditionally_select ... test circuit::ecc::test::test_conditionally_select has been running for over 60 seconds
test circuit::ecc::test::test_edwards_addition ... test circuit::ecc::test::test_edwards_addition has been running for over 60 seconds
test circuit::ecc::test::test_edwards_fixed_base_multiplication ... test circuit::ecc::test::test_edwards_fixed_base_multiplication has been running for over 60 seconds
test circuit::ecc::test::test_edwards_doubling ... test circuit::ecc::test::test_edwards_doubling has been running for over 60 seconds
test circuit::ecc::test::test_edwards_multiplication ... test circuit::ecc::test::test_edwards_multiplication has been running for over 60 seconds
test circuit::ecc::test::test_interpret ... test circuit::ecc::test::test_interpret has been running for over 60 seconds
test circuit::ecc::test::test_into_edwards ... test circuit::ecc::test::test_into_edwards has been running for over 60 seconds
test circuit::ecc::test::test_montgomery_addition ... test circuit::ecc::test::test_montgomery_addition has been running for over 60 seconds
test circuit::multipack::test_multipacking ... test circuit::multipack::test_multipacking has been running for over 60 seconds
test circuit::num::test::test_into_bits ... test circuit::num::test::test_into_bits has been running for over 60 seconds
test circuit::pedersen_hash::test::test_pedersen_hash ... test circuit::pedersen_hash::test::test_pedersen_hash has been running for over 60 seconds
test circuit::pedersen_hash::test::test_pedersen_hash_constraints ... test circuit::pedersen_hash::test::test_pedersen_hash_constraints has been running for over 60 seconds
test circuit::sapling::test_input_circuit_with_bls12_381 ... test circuit::sapling::test_input_circuit_with_bls12_381 has been running for over 60 seconds
test circuit::sapling::test_output_circuit_with_bls12_381 ... test circuit::sapling::test_output_circuit_with_bls12_381 has been running for over 60 seconds
test circuit::sha256::test::test_against_vectors ... test circuit::sha256::test::test_against_vectors has been running for over 60 seconds
test circuit::sprout::test_sprout_constraints ... test circuit::sprout::test_sprout_constraints has been running for over 60 seconds
test jubjub::test_jubjub_bls12 ... test jubjub::test_jubjub_bls12 has been running for over 60 seconds
test redjubjub::tests::cofactor_check ... test redjubjub::tests::cofactor_check has been running for over 60 seconds
test redjubjub::tests::random_signatures ... test redjubjub::tests::random_signatures has been running for over 60 seconds
test redjubjub::tests::round_trip_serialization ... test redjubjub::tests::round_trip_serialization has been running for over 60 seconds
test redjubjub::tests::test_batch_verify ... test redjubjub::tests::test_batch_verify has been running for over 60 seconds
error: process didn't exit successfully: `/mnt/crate/target/debug/deps/fil_sapling_crypto-bc65f9b0eb902842` (signal: 9, SIGKILL: kill)
Exited with code 101

```

arielgabizon commented:
It indeed doesn't seem to make sense to precompute this window 16 table e.g. for a blake2s test;
an easier hack than full configurability might be to just choose between the 8 and 16 case for now..if we're looking for an easy integration with the tests.

porcuquine committed:
Make window_size minimally configurable, with old default.

porcuquine committed:
Use window_size 16 in pedersen tests.

porcuquine commented:
Okay, I made a minimal modification that will allow creation of `JubjubParams` with arbitrary window size, but only when created with `JubjubBls12::new_with_window_size`. This should leave all existing usages unchanged and allows for a simple drop-in replacement wherever we want to use something different.

We can proceed with the work in #736, using the new initialization method. This should still default to 16 and be configurable through setting.

In order to give us a little more assurance, I changed only the pedersen tests to use `window_size` 16.

I also added a benchmark demonstrating that the change works and making it easy to see the speedup:

```
➜  sapling-crypto git:(feat/spedersen-1) cargo bench pedersen
   Compiling fil-sapling-crypto v0.1.1-alpha.0 (/home/porcuquine/dev/sapling-crypto)
    Finished release [optimized] target(s) in 9.97s
     Running target/release/deps/fil_sapling_crypto-a4e1769f0be8e66c

running 2 tests
test circuit::pedersen_hash::test::test_pedersen_hash ... ignored
test circuit::pedersen_hash::test::test_pedersen_hash_constraints ... ignored

test result: ok. 0 passed; 0 failed; 2 ignored; 0 measured; 85 filtered out

     Running target/release/deps/pedersen_hash-7308957cda956ed4

running 2 tests
test bench_pedersen_hash    ... bench:      26,520 ns/iter (+/- 19,501)
test bench_pedersen_hash_16 ... bench:      16,225 ns/iter (+/- 642)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out
```

porcuquine committed:
Add more benchmarks.
